--- callweaver-RC-1.1.99.20070923/channels/misdn/misdn_config.c~	2007-09-23 19:36:25.000000000 +0200
+++ callweaver-RC-1.1.99.20070923/channels/misdn/misdn_config.c	2007-12-30 21:58:16.000000000 +0100
@@ -317,7 +317,7 @@
 		"\t4 - even more Verbose than 3" },
 	{ "misdn_init", MISDN_GEN_MISDN_INIT, MISDN_CTYPE_STR, "/etc/misdn-init.conf", NONE,
 		"Set the path to the misdn-init.conf (for nt_ptp mode checking)." },
-	{ "tracefile", MISDN_GEN_TRACEFILE, MISDN_CTYPE_STR, "/var/log/callweaver.org/misdn.log", NONE,
+	{ "tracefile", MISDN_GEN_TRACEFILE, MISDN_CTYPE_STR, "/var/log/callweaver/misdn.log", NONE,
 		"Set the path to the massively growing trace file, if you want that." },
 	{ "bridging", MISDN_GEN_BRIDGING, MISDN_CTYPE_BOOL, "yes", NONE,
 		"Set this to yes if you want mISDN_dsp to bridge the calls in HW." },
