--- callweaver-1.2.0.1/corelib/dsp.c.org	2008-06-24 21:01:07.696898500 +0000
+++ callweaver-1.2.0.1/corelib/dsp.c	2008-06-24 21:02:33.118237000 +0000
@@ -898,9 +898,9 @@
                                     NULL);
     }
     if ((digitmode & DSP_DIGITMODE_RELAXDTMF))
-        dtmf_rx_parms(&dsp->dtmf_rx, FALSE, 8, 8);
+        dtmf_rx_parms(&dsp->dtmf_rx, FALSE, 8, 8, -99);
     else
-        dtmf_rx_parms(&dsp->dtmf_rx, FALSE, 8, 4);
+        dtmf_rx_parms(&dsp->dtmf_rx, FALSE, 8, 4, -99);
     dsp->digitmode = digitmode;
     return 0;
 }
